ForumProject/notifications/signals.py

```diff
@@ -39,47 +49,83 @@ def create_notifications(instances, follow_status_change=True):
             )
         )
 
-    # Bulk create all notifications
-    if notifications:
-        return Notification.objects.bulk_create(notifications)
-    return []
+    try:
+        # Bulk create all notifications
+        if notifications:
+            return Notification.objects.bulk_create(notifications)
+    except DatabaseError as db_err:
+        logger.error(f"Database error while creating notifications: {str(db_err)}")
 
-# Signal to create a notification when an investor starts following a project or changes a share
-@receiver(post_save, sender=InvestorProject)
-def investor_project_followed_or_subscription_changed(sender, instance, created, **kwargs):
+    return notifications
+
+def do_notifications(instance, follow_status_change=True):
```

Comment:
The complexity in the `do_notifications` method can be reduced by separating concerns more distinctly (e.g., separating email sending logic into a different method)

Reply:
I gave it a thought and must admit I do not know how to implement the proposed concerns separation without adverse effect on the code readability. Quite a big number of arguments would need to be passed to such new _helper_ method designed to separate some concerns `do_notifications` deals with from others.
I also note that the `do_notifications` method already has two _helper_ functions called inside of it - `create_notifications` and `utils.send_email_async`.

Reply:
Let's leave this part without changes. However we'll think about this issue